repo URL: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>/blob/main/slime_sample_notebook.ipynb

commit id: 655d5edf04905edf599edd5b3e338a841bb159de

In [3]:
# Imports section
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import  r2_score
from sklearn.metrics import root_mean_squared_error
from sklearn.model_selection import RepeatedKFold
from sklearn.metrics import make_scorer
from sklearn.pipeline import make_pipeline
# used 383gpt

## Part 1. Loading the dataset

In [4]:
# Using pandas load the dataset (load remotely, not locally)/workspaces/assignment-4-equation-of-a-slime-<redacted>/science_data_large.csv

data = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(f'First 15 rows: \n{data.head(15)}')
# Display a summary of the table information (number of datapoints, etc.)
print(f'Summary of data (rows, columns): \n{data.shape}')
print(f'Summary of data (column names): \n {data.columns}')

First 15 rows: 
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
Summary of data (rows, columns): 
(1000, 3)
Summary of data (column names): 
 Index(['Temperature °C', 'Mols KCL', 'Size nm^3'], dtype='object')


## Part 2. Splitting the dataset

In [9]:
# Take the pandas dataset and split it into our features (X) and label (y)

X = data[['Mols KCL', 'Temperature °C']] 
y = data['Size nm^3']                   

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [10]:
# Use sklearn to train a model on the training set (used 383gpt for this entire section to help me understand)
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample = pd.DataFrame([[500, 600]], columns=['Mols KCL', 'Temperature °C'])
predicted_size = linear_model.predict(sample)
print(f"Predicted size for [500 mols KCl, 600°C] is: {predicted_size[0]:.2f} nm^3")
y_pred = linear_model.predict(X_test)
r2 = r2_score(y_test, y_pred)
rmse = root_mean_squared_error(y_test, y_pred)


print(f"R² score: {r2:.4f}")
print(f"Root mean Squared Error: {rmse:.2f}")
# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = linear_model.coef_
intercept = linear_model.intercept_
print(f'Coefficients: {coefficients}')
print(f'Intercept: {intercept}')

Predicted size for [500 mols KCl, 600°C] is: 626643.90 nm^3
R² score: 0.8552
Root mean Squared Error: 158598.09
Coefficients: [1032.69506649  866.14641337]
Intercept: -409391.4795834082


# Model Output Summary

Results from training the model:

- Predicted size for the sample data point (500 Mols KCl, 600°C): approximately 626,643.90 nm³. This is the model’s estimated size of the slime given these experimental conditions.
  
- Root Mean Squared Error (RMSE): 158,598.09. The RMSE indicates the average error between the model’s predictions and the actual values. A larger value suggests there is still some significant deviation between predictions and actual outcomes.

- R² Score: approximately 0.8552, meaning that 85.52% of the variability in slime size can be explained by the model. This suggests a strong relationship between the independent variables (Mols KCl and Temperature) and the dependent variable (Size).



# Model Equation

The relationship between slime size, temperature, and Mols KCl can be described by the following linear equation:

$
h(x) = -409391.48 + 1032.70 \cdot (\text{Mols KCl}) + 866.15 \cdot (\text{Temperature °C})
$

where:
- Intercept: -409,391.48
- Coefficient for Mols KCl: 1032.70 (this means for each additional mol of KCl, the slime size increases by approximately 1032.70 nm³, holding temperature constant).
- Coefficient for Temperature (°C): 866.15 (this means for each additional degree Celsius, the slime size increases by approximately 866.15 nm³, holding Mols KCl constant).



Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [7]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# RepeatedKFold method
cv = RepeatedKFold(n_splits=5, n_repeats=10, random_state=42)
cv_rmse_scores = cross_val_score(linear_model, X, y, cv=cv, scoring='neg_root_mean_squared_error')
cv_rmse_scores = -cv_rmse_scores

print(f"Cross-Validation RMSE scores: {cv_rmse_scores}")
print(f"Mean RMSE score: {cv_rmse_scores.mean():.4f}")
print(f"Standard Deviation of RMSE scores: {cv_rmse_scores.std():.4f}")


Cross-Validation RMSE scores: [169046.62167316 190096.50742701 153555.59522376 154272.30867299
 168602.99286804 177915.54897828 159259.45275286 165107.59563594
 160761.28030179 173298.21041322 158469.76914323 179384.43748862
 156439.34241182 167553.20832747 176567.44380418 167320.99762814
 175372.11159422 165368.84758398 151137.23817262 176500.65192104
 187884.96389846 150093.33113698 174624.74654405 163392.1806604
 161853.81060864 164320.49059935 159269.11462446 163935.68268874
 173118.29922846 175394.50260915 164198.82462249 170315.36970518
 172277.93788996 166525.22698534 164593.51926116 168776.87220971
 162542.60210201 180545.42974634 160909.40058385 166115.29711674
 159138.7979103  178250.67250384 169105.53096395 160971.16591783
 173406.99492984 184605.25934719 161144.87301076 167786.19807291
 162667.25166028 165271.54689136]
Mean RMSE score: 167581.3211
Standard Deviation of RMSE scores: 8842.3334


# Cross-Validation Results Using Repeated K-Fold

After performing Repeated K-Fold cross-validation with 5 splits and 10 repeats on the linear regression model, I obtained the following RMSE scores for each iteration:

Cross-Validation RMSE scores: [169046.62167316 190096.50742701 153555.59522376 154272.30867299
 168602.99286804 177915.54897828 159259.45275286 165107.59563594
 160761.28030179 173298.21041322 158469.76914323 179384.43748862
 156439.34241182 167553.20832747 176567.44380418 167320.99762814
 175372.11159422 165368.84758398 151137.23817262 176500.65192104
 187884.96389846 150093.33113698 174624.74654405 163392.1806604
 161853.81060864 164320.49059935 159269.11462446 163935.68268874
 173118.29922846 175394.50260915 164198.82462249 170315.36970518
 172277.93788996 166525.22698534 164593.51926116 168776.87220971
 162542.60210201 180545.42974634 160909.40058385 166115.29711674
 159138.7979103  178250.67250384 169105.53096395 160971.16591783
 173406.99492984 184605.25934719 161144.87301076 167786.19807291
 162667.25166028 165271.54689136]

The average RMSE across all folds was approximately 167581.32, with a standard deviation of 8842.33.

The RMSE is a metric that quantifies how well our linear regression model predicts the size of the slime sample based on the input features (Mols KCL and Temperature in C). 

- Mean RMSE: The mean RMSE score of 167581.32 indicates the average magnitude of the model's prediction errors. This value suggests that, on average, the model's predictions can deviate from the actual sizes by about 167,581 nm³, which provides a context for evaluating model performance.

- Standard Deviation: The standard deviation of 8842.33 indicates that there is some variability in the model's performance across different folds. While a smaller standard deviation would indicate more consistent performance, this level of variability suggests that while the model performs reasonably well on average, there are instances where its predictions can vary notably, either positively or negatively.

The combination of the mean and standard deviation RMSE allows us to better understand the reliability of our model. This knowledge is crucial when considering the model for predicting outcomes in real-world scenarios, such as laboratory processing or experimental design.

Theses findings  provide a solid indication that the linear regression model captures the relationship between the independent variables (Mols KCL and Temperature) and the dependent variable (Size nm³ of slime) fairly well. 

However, the model's predictive capability may be influenced by other factors, and further exploration ( feature selection, hyperparameter tuning, or using more complex models) could potentially improve accuracy. 

## Part 5. Using Polynomial Regression

In [8]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
model = make_pipeline(PolynomialFeatures(degree=2), LinearRegression())
model.fit(X_train, y_train)
y_pred_poly = model.predict(X_test)
rmse_poly = np.sqrt(root_mean_squared_error(y_test, y_pred_poly))

print(f'Polynomial Regression RMSE: {rmse_poly:.4f}')

linear_regressor = model.named_steps['linearregression']
coefficients = linear_regressor.coef_
intercept = linear_regressor.intercept_
print(f'Coefficients: {coefficients}')
print(f'Intercept: {intercept}')

# Report on the metrics and output the resultant equation as you did in Part 3.

Polynomial Regression RMSE: 0.0110
Coefficients: [ 0.00000000e+00 -1.27196088e-07  1.20000000e+01  2.85714287e-02
  2.00000000e+00  1.26471056e-11]
Intercept: 2.0477862562984228e-05


# Polynomial Regression Equation

Polynomial Regression RMSE: 0.0110

The polynomial regression performed much better at capturing the non-linear relationships between the input variables and the size of the slime, as reflected in the RMSE (lower score).

The polynomial regression model can be expressed as:

$
h(x) = \text{intercept} + c_1 \cdot (\text{Mols KCL}) + c_2 \cdot (\text{Temperature °C}) + c_3 \cdot (\text{Mols KCL})^2 + c_4 \cdot (\text{Temperature °C})^2 + c_5 \cdot (\text{Mols KCL} \cdot \text{Temperature °C})
$

Where:
- Intercept = 0.0110
- Coefficients:
  - $ c_1 = -1.27196088 \times 10^{-7}$ (coefficient for Mols KCL)
  - $c_2 = 12.0$ (coefficient for Temperature °C)
  - $ c_3 = 0.0285714287$ (coefficient for {Mols KCL}^2)
  - $ c_4 = 2.0$ (coefficient for {Temperature °C}^2)
  - $c_5 = 1.26471056 \times 10^{-11}$ (coefficient for {Mols KCL} $\cdot$ {Temperature °C})